js/ui/appIconBar.js

```diff
@@ -631,21 +641,52 @@ const ScrolledIconList = GObject.registerClass({
 
         appSys.connect('installed-changed', this._onInstalledChanged.bind(this));
         appSys.connect('app-state-changed', this._onAppStateChanged.bind(this));
+
+        this._parentalControlsManager.connect('changed', () => {
+            for (let [app, appButton] of this._taskbarApps) {
+                if (app.state == Shell.AppState.STOPPED &&
+                    !this._parentalControlsManager.shouldShowApp(app.get_app_info()))
+                    appButton.hide();
+                else
+                    appButton.show();
+            }
+        });
     }
 
     setActiveApp(app) {
         this._taskbarApps.forEach((appButton, taskbarApp) => {
-            if (app == taskbarApp)
-                appButton.add_style_pseudo_class('highlighted');
-            else
-                appButton.remove_style_pseudo_class('highlighted');
+            var changed = false;
+            var isHighlighted = appButton.has_style_pseudo_class('highlighted');
+
+            if (app == taskbarApp) {
+                if (!isHighlighted) {
+                    appButton.add_style_pseudo_class('highlighted');
+                    changed = true;
+                }
+            } else {
+                if (isHighlighted) {
+                    appButton.remove_style_pseudo_class('highlighted');
+                    changed = true;
+                }
+            }
+            if (changed)
+                appButton.queue_redraw();
         });
     }
 
     getNumAppButtons() {
         return this._taskbarApps.size;
     }
 
+    getNumVisibleAppButtons() {
+        var count = 0;
+        this._taskbarApps.forEach((appButton, app) => {
```

Comment:
This function can be simplified to:

```javascript
return this._taskbarApps.reduce((counter, appButton, index, array) => {
    return appButton.visible ? counter : counter + 1;
}, 0);
```

Reply:
Done - had to do it a bit differently as `this._taskbarApps` is a map (not an array), but still used `Array.reduce`.